Formula/easy-rsa.rb

```diff
@@ -0,0 +1,61 @@
+class EasyRsa < Formula
+  desc "CLI utility to build and manage a PKI CA"
+  homepage "https://github.com/OpenVPN/easy-rsa"
+  url "https://github.com/OpenVPN/easy-rsa/archive/v3.0.8.tar.gz"
+  sha256 "fd6b67d867c3b8afd53efa2ca015477f6658a02323e1799432083472ac0dd200"
+  license "GPL-2.0-only"
+  head "https://github.com/OpenVPN/easy-rsa.git"
+
+  depends_on "openssl@1.1"
+
+  def install
+    # Rename and edit vars.example file to define easy-rsa variables for homebrew
```

Comment:
Are these the only variables? If so, just use `(share"/easyrsa3/vars").write`. If not, this needs to be a post-install block that checks if such a file already exists.

Reply:
The file [`vars.example`](https://github.com/OpenVPN/easy-rsa/blob/master/easyrsa3/vars.example) comes with the package and must be renamed to `vars` and edited to configure it, and the environment variable `$EASY_RSA` needs to be set to its location.

The only other option I can see would involve instructing the user to [set the other two as environment variables](https://github.com/OpenVPN/easy-rsa/blob/master/doc/EasyRSA-Advanced.md#environmental-variables-reference) as well, rather than editing the default config file. I am not sure which is more consistent with Homebrew's philosophy. Maybe that would be preferable, since we have to have the user set `$EASYRSA` either way?

Reply:
I think a wrapper using `write_env_script` would be a better idea then. You can already set the variables that way.

Reply:
Is there some documentation or an example you can point me to? I'm looking at this [formula for `yarn`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/yarn.rb), but to be honest, I'm not sure what's happening here or how it works exactly. And [this documentation](https://www.rubydoc.info/github/Homebrew/brew/Pathname:write_env_script) isn't super-helpful. I gather that this would allow the formula to create a shell script that would set these three variables? And then do we tell the user to source that in `~/.bash_profile`/`~/.zshrc` instead of the individual `export`s, or does that kind of happen by some other magic then?

Reply:
What would happen is that you install the main executable to `#{libexec}` and do 
```ruby
(bin/"executable").write_env_script "#{libexec}/main_file",
      VAR1: "something",
      VAR2: "stuff"
```

Then, `executable` will be the thing the users execute and it'll have all the environment variables set for them in advance since `executable` will be a shell script setting them.

Reply:
Okay, I think I've got it figured out. That's pretty slick. (As an aside, should most other formulae that direct the user to set their own environment variables in the `caveats` block actually be modified to use this method instead?)

My only concern is that, according to the `easy-rsa` docs, [environment variables will override settings in `vars`](https://github.com/OpenVPN/easy-rsa/blob/master/doc/EasyRSA-Advanced.md#configuration-reference) if provided. By using `write_env_script` we are effectively hard-coding these options for the user, and they will have no way—other than command line switches—to change them. `$EASYRSA` obviously should be specified in this manner, since it tells the program where to locate `vars` and other configuration files. `$EASYRSA_OPENSSL` maybe should be, but only if we want to force the user to always use Homebrew's `openssl`. However, `$EASYRSA_PKI` probably shouldn't be, because it would prevent the user from ever changing the default key location (although they could still do it on a case-by-case basis with command line switches).

I'm thinking that `$EASYRSA` should be set through `write_env_script`, but that we should continue to set `$EASYRSA_OPENSSL` and `$EASYRSA_PKI` by modifying `vars.example` and installing it as `vars`, which preserve's the user's ability to specify their own defaults as necessary.

Still to do:
- [ ] figure out how to check for #{opt_share}/vars during install and skip copying if it exists, so we don't clobber those potential user-modified customizations during an upgrade

Reply:
I'd like to make it so that Homebrew doesn't overwrite `#{share}/"vars"` when upgrading, if it exists. 

I tried this:

```
    unless (share/"vars").exist?
      inreplace "easyrsa3/vars.example" do |s|
        s.gsub! "#set_var EASYRSA\t\"${0%/*}\"",
                "#set_var EASYRSA\t\"#{opt_share}\""
        s.gsub! "#set_var EASYRSA_OPENSSL\t\"openssl\"",
                "set_var EASYRSA_OPENSSL\t\"#{Formula["openssl@1.1"].bin}/openssl\""
        s.gsub! "#set_var EASYRSA_PKI\t\t\"$PWD/pki\"",
                "set_var EASYRSA_PKI\t\t\"#{etc}/pki\""
      end
      share.install "easyrsa3/vars.example" => "vars"
    end
```
and then installed 3.0.7 followed by an upgrade to 3.0.8, but it doesn't seem to work, as it overwrote the file anyway. (Maybe an upgrade is just an uninstall and then a fresh install of the new version?) Is there an official way to keep config files across upgrades? Would it require storing it out of Homebrew's namespace, e.g. in `~/.easy-rsa/vars`?

Reply:
> Okay, I think I've got it figured out. That's pretty slick. (As an aside, should most other formulae that direct the user to set their own environment variables in the caveats block actually be modified to use this method instead?)

Only if it's always going to be pointing to homebrew values.

> $EASYRSA_OPENSSL maybe should be, but only if we want to force the user to always use Homebrew's openssl

We do

>  However, $EASYRSA_PKI probably shouldn't be, because it would prevent the user from ever changing the default key location (although they could still do it on a case-by-case basis with command line switches).

https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Shell-Parameter-Expansion for things like `JAVA_HOME` that are sometimes overriden we use this

> I'd like to make it so that Homebrew doesn't overwrite #{share}/"vars" when upgrading, if it exists.

You need to add a `post_install` block. Check `httpd` or `mysql` for examples

Reply:
It dawned on me that the config files should probably be living in `/usr/local/etc/easy-rsa` rather than in the `share` directory of the individual version installation. That solves the persistence across upgrades and uninstallations problem by virtue of the way Homebrew protects those files.

I moved all the config options back into the environment as you suggested, using `.write_env_script` and permitting `$EASYRSA_PKI` to be overridden with shell parameter expansion. `vars.example` is no longer modified on installation, but rather simply copied into `/usr/local/etc/easy-rsa` alongside the other configuration files, where the user could rename and edit it to further configure the package.

I think that cleans up all the issues, unless I've created new issues. 😧